### PR TITLE
[RLlib] POC: Separate losses for PPO/APPO/IMPALA. Enable TFPolicy to handle multiple optimizers/losses (like TorchPolicy).

### PR DESCRIPTION
**Pull request by sven1977**
POC:
- Separate losses for APPO/IMPALA.
- Enable TFPolicy to handle multiple optimizers/losses (like TorchPolicy already does).

An experimental flag: `_tf_policy_handles_more_than_one_loss` must be set to True to allow for TFPolicy to handle more than one loss term returned by the loss function. In this case, the same number of local TF optimizers must be returned by the policy's `optimizer` (arg `optimizer_fn` for `build_tf_policy()`).
This is so that the user has to explicitly set this flag to activate this change in the TFPolicy API.

To use this feature for APPO or IMPALA:
- Set "_separate_vf_optimizer" to True
- Make sure the model has completely separate policy- and value branches.
- Set the learning rate for the value function's optimizer via `_lr_vf`.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/tf_policy.py [hunk 1]
```diff
@@ -330,18 +342,35 @@ def _initialize_loss(self, loss: TensorType,
...
+            self._optimizer = self._optimizers[0] if self._optimizers else None
+
+        # Supporting more than one loss/optimizer.
+        if self.config["_tf_policy_handles_more_than_one_loss"]:
+            self._grads_and_vars = []
+            self._grads = []
+            for group in self.gradients(self._optimizers, self._losses):
+                g_and_v = [(g, v) for (g, v) in group if g is not None]
+                self._grads_and_vars.append(g_and_v)
+                self._grads.append([g for (g, v) in g_and_v])
```
Comment (gjoliver) [inline rllib/policy/tf_policy.py:365]: don't know if our linter will catch unused variable. since v is not used here:

[g for (g, _) in g_nd_v]

same below.

File: rllib/policy/tf_policy.py [hunk 2]
```diff
@@ -330,18 +342,35 @@ def _initialize_loss(self, loss: TensorType,
...
+        # Only one optimizer and and loss term.
+        else:
+            self._grads_and_vars = [
+                (g, v)
+                for (g, v) in self.gradients(self._optimizer, self._loss)
```
Comment (gjoliver) [inline rllib/policy/tf_policy.py:370]: after seeing all these ifs, I wonder if the code will be simpler if we always use self._optimizers and self._losses here.
and handle things as single element array everywhere.

File: rllib/policy/tf_policy.py [hunk 3]
```diff
@@ -245,22 +248,30 @@ def __init__(self,
...
-        self._optimizer_variables = None
+        self._optimizer_variables: \
+            Optional[ray.experimental.tf_utils.TensorFlowVariables] = None
 
-        # The loss tf-op.
+        # The loss tf-op(s). Number of losses must match number of optimizers.
+        self._losses = []
+        # Backward compatibility (in case custom child TFPolicies access this
+        # property).
         self._loss = None
```
Comment (gjoliver) [inline rllib/policy/tf_policy.py:269]: do we have to keep self._loss here, instead of moving everything to use self._losses, because of backward compatibility reasons?
like, someone may have been overriding things with their custom tf policy class?

I thought this is DeveloperAPI, so we should be able to move everything to self._optimizers and self._losses?

**Review by gjoliver — COMMENTED** [review]
A lot of questions.
In general, I wonder if we should get rid of self._optimizer and self._loss, and move to self._optimizers and self._losses everywhere, so we don't get so many if branches.

File: rllib/agents/impala/impala.py [hunk 4]
```diff
@@ -230,6 +236,23 @@ def validate_config(config):
...
+    # `_tf_policy_handles_more_than_one_loss` to True.
+    if config["_separate_vf_optimizer"] is True:
+        # Only supported to tf so far.
+        # TODO(sven): Need to change APPO|IMPALATorchPolicies (and the models
+        #  to return separate sets of weights in order to create the different
+        #  torch optimizers).
+        if config["framework"] not in ["tf", "tf2", "tfe"]:
+            raise ValueError(
+                "`_separate_vf_optimizer` only supported to tf so far!")
+        if config["_tf_policy_handles_more_than_one_loss"] is False:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/impala/impala.py:249]: "is there ever a case where we have 1 optimizer but multiple loss function?"
No, but this is an existing API that we would change if we didn't carefully introduced this new flag. The user has to explicitly set this flag to True to agree to the API change. Otherwise, we would break code for existing users of our TFPolicy API.
I think this is a typical path we should go whenever we change an API:
1) Flag explicitly set to allow for the new API to be used.
2) Set this flag to True by default, catch possible errors for those users that are still on the old API, and warn if flag is left False.
3) Error if flag is left to False.
4) Remove flag and all checks for it in the code: Behave always as if flag was set to True.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/impala/impala.py:249]: No, this has nothing to do with the model builder API. It's a plain change in the TFPolicy API. Note also that the TorchPolicy API is already behaving like this (flag is True): You can return more than one loss term from the loss function, but then also need the same number of optimizers returned from your optimizer function.

So we are just unifying TFPolicy and TorchPolicy behavior (but we have to do this carefully as to not break existing code!).

File: rllib/agents/impala/vtrace_tf_policy.py [hunk 5]
```diff
@@ -251,27 +252,62 @@ def grad_stats(policy, train_batch, grads):
...
+            optim = tf.keras.optimizers.Adam(policy.cur_lr)
+            if policy.config["_separate_vf_optimizer"]:
+                return optim, tf.keras.optimizers.Adam(policy.config["_lr_vf"])
         else:
-            return tf1.train.AdamOptimizer(policy.cur_lr)
+            optim = tf1.train.AdamOptimizer(policy.cur_lr)
+            if policy.config["_separate_vf_optimizer"]:
+                return optim, tf1.train.AdamOptimizer(policy.config["_lr_vf"])
     else:
+        if policy.config["_separate_vf_optimizer"]:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/impala/vtrace_tf_policy.py:263]: Yeah, agreed, this is messy. We should unify the optimizer configurations. Right now, most algos create Adam by default (and users can't really do anything about this, unless they override the `optimizer_fn` entirely).

I think this is a separate problem from the new flags, though.

It's two different (experimental) flags we introduce here:
1) Allow TFPolicy to behave more like TorchPolicy in its optimizer/loss API.
2) Use two losses/optimizers for APPO/IMPALA (which - if True - requires (1) ).

File: rllib/agents/impala/vtrace_tf_policy.py [hunk 6]
```diff
@@ -251,27 +252,62 @@ def grad_stats(policy, train_batch, grads):
...
+            clipped_g_and_v = []
+            for g, v in grads_and_vars:
+                if g is not None:
+                    clipped_g, _ = tf.clip_by_global_norm(
+                        [g], policy.config["grad_clip"])
+                    clipped_g_and_v.append((clipped_g[0], v))
+            clipped_grads_and_vars.append(clipped_g_and_v)
+
+        policy.grads = [
+            g for g_and_v in clipped_grads_and_vars for (g, v) in g_and_v
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/impala/vtrace_tf_policy.py:298]: It does :) , I mean work.
Yeah, now with >1 optimizers, your grads_and_vars is a `List[List[ModelGradients]]`, whereas with the old API, it's a `List[ModelGradients]`.

File: rllib/agents/impala/vtrace_tf_policy.py [hunk 7]
```diff
@@ -251,27 +252,62 @@ def grad_stats(policy, train_batch, grads):
...
-    return clipped_grads
+    # Supporting more than one loss/optimizer.
+    if policy.config["_tf_policy_handles_more_than_one_loss"]:
+        optimizers = force_list(optimizer)
+        losses = force_list(loss)
+        assert len(optimizers) == len(losses)
+        clipped_grads_and_vars = []
+        for optim, loss_ in zip(optimizers, losses):
+            grads_and_vars = optim.compute_gradients(
+                loss_, policy.model.trainable_variables())
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/impala/vtrace_tf_policy.py:288]: Correct, tf is good at distinguishing these. So it's ok to pass in all the model's variables, even though some of these vars won't have gradients under the given loss term (their values will be None).

File: rllib/policy/dynamic_tf_policy.py [hunk 8]
```diff
@@ -812,28 +818,62 @@ def __init__(
...
+                        clipped.append(grad)
+                    clipped, _ = tf.clip_by_global_norm(
+                        clipped, grad_norm_clipping)
+                    for i, (grad, var) in enumerate(avg):
+                        avg[i] = (clipped[i], var)
+                avgs.append(avg)
+
+            # Gather update ops for any batch norm layers.
+            # TODO(ekl) here we
+            #  will se all the ops found which won't work for DQN / DDPG, but
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/policy/dynamic_tf_policy.py:837]: +1

File: rllib/policy/tf_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/policy/tf_policy.py:365]: +1

File: rllib/policy/tf_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/policy/tf_policy.py:370]: Normally, yes, but the problem is those children of TFPolicy that don't have the `self._optimizers` property as they inherited directly from TFPolicy and did not build the class with the `build_tf_policy` utility function. :)

File: rllib/policy/tf_policy.py (hunk 3 above)
Comment (sven1977) [reply rllib/policy/tf_policy.py:269]: Correct, lots of users probably still directly access `self._loss` or `self._optimizer` in their TFPolicy child classes.

File: rllib/policy/dynamic_tf_policy.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/policy/dynamic_tf_policy.py:837]: done

File: rllib/policy/tf_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/policy/tf_policy.py:365]: done

File: rllib/policy/tf_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/policy/tf_policy.py:370]: We are (very carefully) introducing a new API here and have to make 100% sure, nothing breaks for users (and some of our algos) still using the old one. Once we can deprecate the old API, we can remove all these ugly if-blocks. But that'll take some time and some PRs introducing soft-warnings for users still on the old API.

File: rllib/agents/ppo/appo_tf_policy.py [hunk 9]
```diff
@@ -250,24 +250,32 @@ def reduce_mean_valid(t):
...
     policy._mean_policy_loss = mean_policy_loss
     policy._mean_kl = mean_kl
     policy._mean_vf_loss = mean_vf_loss
     policy._mean_entropy = mean_entropy
     policy._value_targets = value_targets
 
-    # Store stats in policy for stats_fn.
-    return total_loss
+    # Return one total loss or two losses: vf vs rest (policy + kl).
+    if policy.config["_separate_vf_optimizer"]:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/ppo/appo_tf_policy.py:275]: Not sure, there is also the entropy_loss, which would then be in the `policy_loss`?
I think it's more expressive here to use these terms.
total_loss: Makes it clear that this is one single loss containing all sub-terms.
loss_wo_vf: The loss, but w/o the value function term.

I also changed `mean_kl` into `mean_kl_loss` (backward compatibly ;) ) to match the other loss names

**Review by gjoliver — APPROVED** [review]
ok, I think the PR makes much more sense now that you mentioned the flag guarded dev process.
looks great to me then.

feel like we should try very hard to consolidate things, and get rid of these dev flags over time.

**Comment by gjoliver** [comment]
there may still be a broken unit test though.
